src/plugins/upf/flowtable_init.c

```diff
@@ -108,13 +112,39 @@ flowtable_init_cpu (flowtable_main_t * fm, flowtable_main_per_cpu_t * fmt)
   return error;
 }
 
+static void
+flowtable_gauge_update_flow_size (stat_segment_directory_entry_t * e, u32 i)
+{
+  flowtable_main_t *fm = &flowtable_main;
+  vlib_thread_main_t *tm = vlib_get_thread_main ();
+  u32 cpu_index;
+  flowtable_main_per_cpu_t *fmt;
+
+  if (!fm)
+    return;
+
+  e->value = 0;
+
+  vec_validate (fm->per_cpu, tm->n_vlib_mains - 1);
+  for (cpu_index = 0; cpu_index < tm->n_vlib_mains; cpu_index++) {
+      // TODO: we are not thread-aware here. Add thread awarness / locks
+      // in case something goes wrong with workers count > 1
+      fmt = &fm->per_cpu[cpu_index];
+      if (!fmt)
+        continue;
+      e->value += fmt->flows_cnt;
+  }
+
+}
```

Comment:
this (and the rest of the flows_cnt) basically duplicated vlib_get_simple_counter and friends. Why don't you simply use what is already there?

Reply:
from what I've seen so far, in VPP world counter is something that should be only incremented(friends are quite around it). I've tried to avoid unreadable logic around simple counters just to have an ability to decrement a counter

Reply:
Would it be possible to use the simple counters and implement the decrement method only? It should just be a matter of copying the increment and change the operation, right?

Reply:
Yes sure. And it's worth to be an upstream change also.
Another possible thing is to fetch counter value, dec it and then set it back. However option with new method is better

Reply:
while the upstream discussion about gauges is going on, at least fix this method.

1. the vec_validate is not needed,
2. just iterate of the vector (vec_foreach or simple loop over the index up to vec_len), there is no need to involve the tm->n_vlib_mains
3. the `if (!fmt)` check is wrong, fmt is guaranteed not to be NULL, *fmt can only be NULL is the init fails, but that should abort VPP

Reply:
Fixed